### PR TITLE
Allow package authors to provide a custom error message when an uninitialized field is accessed

### PR DESCRIPTION
**Pull request by DilumAluthge**
## Example

The package author would have code like this in their package:

```julia
module MyPackage

using LazilyInitializedFields: @lazy, uninit, UninitializedFieldException

@lazy struct Foo
    @lazy a::Int
    @lazy b::Int
end

Foo() = Foo(uninit, uninit)

function __init__()
    if isdefined(Base.Experimental, :register_error_hint)
        Base.Experimental.register_error_hint(UninitializedFieldException) do io, exc
            if exc.T === Foo
                if exc.s === :a
                    print(io, "\nThis is a custom hint for the `:a` field of the `Foo` struct.")
                elseif exc.s === :b
                    print(io, "\nHere's my custom hint for the `:a` field of the `Foo` struct.")
                end
            end
        end
    end
end

end # module
```

And then a package user would see something like this:

```julia-repl
julia> using .MyPackage: Foo

julia> f = Foo()
Foo(uninit, uninit)

julia> f.a
ERROR: field `a` in struct of type `Foo` is not initialized
This is a custom hint for the `:a` field of the `Foo` struct.
Stacktrace:
 [1] getproperty(x::Foo, s::Symbol)
   @ Main.MyPackage ~/Downloads/LazilyInitializedFields.jl/src/LazilyInitializedFields.jl:316
 [2] top-level scope
   @ REPL[4]:1

julia> f.b
ERROR: field `b` in struct of type `Foo` is not initialized
Here's my custom hint for the `:a` field of the `Foo` struct.
Stacktrace:
 [1] getproperty(x::Foo, s::Symbol)
   @ Main.MyPackage ~/Downloads/LazilyInitializedFields.jl/src/LazilyInitializedFields.jl:316
 [2] top-level scope
   @ REPL[5]:1
```

### CODE REVIEW

**Comment by DilumAluthge** [comment]
cc: @andreasnoack

**Comment by DilumAluthge** [comment]
Bump @KristofferC @andreasnoack

**Comment by DilumAluthge** [comment]
I've switched the implementation of this PR to use the `Base.Experimental.register_error_hint` functionality instead.

**Comment by andreasnoack** [comment]
I'm not really sure what the better approach here is but having some way of providing a more useful error message would help our use case a lot. Users of the libraries that we develop won't really know what to do differently if they are just told that a field is uninitialized, but we can in most cases provide reasonable hints based on the context. We just need a way to provide them.

File: src/LazilyInitializedFields.jl [hunk 1]
```diff
@@ -93,8 +93,12 @@ struct UninitializedFieldException <: Exception
     T::DataType
     s::Symbol
 end
-Base.showerror(io::IO, err::UninitializedFieldException) =
+function Base.showerror(io::IO, err::UninitializedFieldException)
     print(io, "field `", err.s, "` in struct of type `$(err.T)` is not initialized")
+    if isdefined(Base.Experimental, :show_error_hints)
```
Comment (KristofferC) [inline src/LazilyInitializedFields.jl]: Should this use `@static`?

**Comment by KristofferC** [comment]
Seems reasonable to me, would be good with a test though.

File: src/LazilyInitializedFields.jl (hunk 1 above)
Comment (DilumAluthge) [reply src/LazilyInitializedFields.jl]: Yes. Done!

**Comment by DilumAluthge** [comment]
I've added some tests!

It would probably be good if you could first merge #11. Then I can rebase this PR, and we can make sure that CI passes on this PR before you merge it.

**Comment by KristofferC** [comment]
Could you please just inline the tests into the `runtests.jl` file. I don't really see a point with splitting it up.

**Comment by DilumAluthge** [comment]
Done!

**Comment by DilumAluthge** [comment]
Bump @KristofferC

**Comment by DilumAluthge** [comment]
Thanks!

**Pull request merged**